app/models/container_deployment_node.rb

```diff
@@ -4,4 +4,41 @@ class ContainerDeploymentNode < ApplicationRecord
   serialize :labels, Hash
   serialize :customizations, Hash
   acts_as_miq_taggable
+
+  def node_address
+    if vm
+      vm.hostnames.last || vm.hardware.ipaddresses.last
+    elsif address
+      address
+    end
+  end
+
+  def labels_hash
+    labels ||= {}
+    labels.each do |key, val|
```

Comment:
Is labels not already a hash (so you can just .dup)?  If not, this whole method might be simpler to write as `Hash[labels.to_a]`